File: 백준/Silver/10845. 큐/큐.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Queue {
    int front, rear;
    int *arr;
} Queue;

Queue *createQueue(int n) {
    Queue *q = (Queue *)malloc(sizeof(Queue));
    q->front = q->rear = 0;
    q->arr = (int *)malloc((n + 1) * sizeof(int));
    return q;
}

int isFull(Queue *q, int n) {
    return (q->rear + 1) % (n + 1) == q->front;
}

int isEmpty(Queue *q) {
    return q->front == q->rear;
}

void enQueue(Queue *q, int data, int n) {
    if(isFull(q, n)) {
        return;
    }
    q->rear = (q->rear + 1) % (n + 1);
    q->arr[q->rear] = data;
}

int deQueue(Queue *q, int n) {
    if(isEmpty(q)) {
        return -1;
    }
    q->front = (q->front + 1) % (n + 1);
    return q->arr[q->front];
}

int checkSize(Queue *q, int n) {
    return (q->rear - q->front + (n + 1)) % (n + 1);
}

int getFront(Queue *q, int n) {
    if(isEmpty(q))
        return -1;
    return q->arr[(q->front + 1) % (n + 1)];
}

int getBack(Queue *q, int n) {
    if(isEmpty(q))
        return -1;
    return q->arr[q->rear];
}

int main() {
    int n;
    scanf("%d", &n);
    
    Queue *q = createQueue(n);
    for (int i = 0; i < n; i++) {
        char cmd[10];
        scanf("%s", cmd);
        
        if (strcmp(cmd, "push") == 0) {
            int data;
            scanf("%d", &data);
            enQueue(q, data, n);
        }
        else if (strcmp(cmd, "pop") == 0) {
            printf("%d\n", deQueue(q, n));
        }
        else if (strcmp(cmd, "size") == 0) {
            printf("%d\n", checkSize(q, n));
        }
        else if (strcmp(cmd, "empty") == 0) {
            printf("%d\n", isEmpty(q) ? 1 : 0);
        }
        else if (strcmp(cmd, "front") == 0) {
            printf("%d\n", getFront(q, n));
        }
        else if (strcmp(cmd, "back") == 0) {
            printf("%d\n", getBack(q, n));
        }
    }
    
    free(q->arr);
    free(q);
    return 0;
}

File: 백준/Silver/10845. 큐/큐.py
# 정수를 저장하는 큐 구현 
# push X: 정수 X를 큐에 넣는 연산이다.
# pop: 큐에서 가장 앞에 있는 정수를 빼고, 그 수를 출력한다. 만약 큐에 들어있는 정수가 없는 경우에는 -1을 출력한다.
# size: 큐에 들어있는 정수의 개수를 출력한다.
# empty: 큐가 비어있으면 1, 아니면 0을 출력한다.
# front: 큐의 가장 앞에 있는 정수를 출력한다. 만약 큐에 들어있는 정수가 없는 경우에는 -1을 출력한다.
# back: 큐의 가장 뒤에 있는 정수를 출력한다. 만약 큐에 들어있는 정수가 없는 경우에는 -1을 출력한다.

import sys

input = sys.stdin.readline

N = int(input())
queue = []

for _ in range(N):
    cmd = input().split()
    if cmd[0] == 'push':
        queue.append(cmd[1])
    elif cmd[0] == 'pop':
        if queue:
            print(queue.pop(0))
        else:
            print(-1)
    elif cmd[0] == 'size':
        print(len(queue))
    elif cmd[0] == 'empty':
        print(0 if queue else 1)
    elif cmd[0] == 'front':
        print(queue[0] if queue else -1)
    elif cmd[0] == 'back':
        print(queue[-1] if queue else -1)
    else:
        pass